backend/src/schema/resolvers/emails.spec.js

```diff
@@ -0,0 +1,298 @@
+                  verifiedAt: expect.any(String),
+                  createdAt: expect.any(String),
+                },
+              },
+              errors: undefined,
+            })
+          })
+
+          it('connects the new `EmailAddress` as PRIMARY', async () => {
+            await mutate({ mutation, variables })
+            const result = await neode.cypher(`
+            MATCH(u:User {id: "567"})-[:PRIMARY_EMAIL]->(e:EmailAddress {email: "to-be-verified@example.org"})
+            RETURN e
+          `)
+            const email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email.toJson()).resolves.toMatchObject({
+              email: 'to-be-verified@example.org',
+            })
+          })
+
+          it('removes previous PRIMARY relationship', async () => {
+            const cypherStatement = `
+            MATCH(u:User {id: "567"})-[:PRIMARY_EMAIL]->(e:EmailAddress {email: "user@example.org"})
+            RETURN e
+          `
+            let result = await neode.cypher(cypherStatement)
+            let email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email.toJson()).resolves.toMatchObject({
+              email: 'user@example.org',
+            })
+            await mutate({ mutation, variables })
+            result = await neode.cypher(cypherStatement)
+            email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email).toBe(false)
+          })
+
+          it('removes previous `EmailAddress` node', async () => {
+            const cypherStatement = `
+            MATCH(u:User {id: "567"})<-[:BELONGS_TO]-(e:EmailAddress {email: "user@example.org"})
+            RETURN e
+          `
+            let result = await neode.cypher(cypherStatement)
+            let email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email.toJson()).resolves.toMatchObject({
+              email: 'user@example.org',
+            })
+            await mutate({ mutation, variables })
+            result = await neode.cypher(cypherStatement)
+            email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email).toBe(false)
+          })
+
+          describe('Edge case: In the meantime someone created an `EmailAddress` node with the given email', () => {
+            beforeEach(async () => {
+              await factory.create('EmailAddress', { email: 'to-be-verified@example.org' })
+            })
+
+            it('throws UserInputError because of unique constraints', async () => {
+              await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+                data: { VerifyEmailAddress: null },
+                errors: [{ message: 'A user account with this email already exists.' }],
+              })
+            })
+          })
+        })
+      })
+    })
+  })
+})
```

Comment:
> [<img alt="alina-beck" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/alina-beck) **Authored by [alina-beck](https://github.com/alina-beck)**
_<time datetime="2019-09-30T08:49:59Z" title="Monday, September 30th 2019, 10:49:59 am +02:00">Sep 30, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,283 @@
+import Factory from '../../seed/factories'
+import { gql } from '../../jest/helpers'
+import { getDriver, neode as getNeode } from '../../bootstrap/neo4j'
+import createServer from '../../server'
+import { createTestClient } from 'apollo-server-testing'
+
+const factory = Factory()
+const neode = getNeode()
+
+let mutate
+let authenticatedUser
+let user
+let variables
+const driver = getDriver()
+
+beforeEach(async () => {
+  variables = {}
+})
+
+beforeAll(() => {
+  const { server } = createServer({
+    context: () => {
+      return {
+        driver,
+        neode,
+        user: authenticatedUser,
+      }
+    },
+  })
+  mutate = createTestClient(server).mutate
+})
+
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
+
+describe('AddEmailAddress', () => {
+  const mutation = gql`
+    mutation($email: String!) {
+      AddEmailAddress(email: $email) {
+        email
+        verifiedAt
+        createdAt
+      }
+    }
+  `
+  beforeEach(() => {
+    variables = { ...variables, email: 'new-email@example.org' }
+  })
+
+  describe('unauthenticated', () => {
+    beforeEach(() => {
+      authenticatedUser = null
+    })
+
+    it('throws AuthorizationError', async () => {
+      await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+        data: { AddEmailAddress: null },
+        errors: [{ message: 'Not Authorised!' }],
+      })
+    })
+  })
+
+  describe('authenticated', () => {
+    beforeEach(async () => {
+      user = await factory.create('User', { email: 'user@example.org' })
+      authenticatedUser = await user.toJson()
+    })
+
+    describe('email attribute is not a valid email', () => {
+      beforeEach(() => {
+        variables = { ...variables, email: 'foobar' }
+      })
+
+      it('throws UserInputError', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: { AddEmailAddress: null },
+          errors: [{ message: 'must be a valid email' }],
+        })
+      })
+    })
+
+    describe('email attribute is a valid email', () => {
+      it('creates a new unverified `EmailAddress` node', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: {
+            AddEmailAddress: {
+              email: 'new-email@example.org',
+              verifiedAt: null,
+              createdAt: expect.any(String),
+            },
+          },
+          errors: undefined,
+        })
+      })
+
+      it('connects `EmailAddressRequest` to the authenticated user', async () => {
+        await mutate({ mutation, variables })
+        const result = await neode.cypher(`
+        MATCH(u:User)-[:PRIMARY_EMAIL]->(:EmailAddress {email: "user@example.org"})
+        MATCH(u:User)<-[:BELONGS_TO]-(e:EmailAddressRequest {email: "new-email@example.org"})
+        RETURN e
+      `)
+        const email = neode.hydrateFirst(result, 'e', neode.model('EmailAddressRequest'))
+        await expect(email.toJson()).resolves.toMatchObject({
+          email: 'new-email@example.org',
+          nonce: expect.any(String),
+        })
+      })
+
+      describe('if another `EmailAddressRequest` node already exists with that email', () => {
+        it('throws no unique constraint violation error', async () => {
+          await factory.create('EmailAddressRequest', {
+            createdAt: '2019-09-24T14:00:01.565Z',
+            email: 'new-email@example.org',
+          })
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: {
+              AddEmailAddress: {
+                email: 'new-email@example.org',
+                verifiedAt: null,
+              },
+            },
+            errors: undefined,
+          })
+        })
+      })
+
+      describe('but if another user owns an `EmailAddress` already with that email', () => {
+        it('throws UserInputError because of unique constraints', async () => {
+          await factory.create('User', { email: 'new-email@example.org' })
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: { AddEmailAddress: null },
+            errors: [{ message: 'A user account with this email already exists.' }],
+          })
+        })
+      })
+    })
+  })
+})
+
+describe('VerifyEmailAddress', () => {
+  const mutation = gql`
+    mutation($email: String!, $nonce: String!) {
+      VerifyEmailAddress(email: $email, nonce: $nonce) {
+        email
+        createdAt
+        verifiedAt
+      }
+    }
+  `
+
+  beforeEach(() => {
+    variables = { ...variables, email: 'to-be-verified@example.org', nonce: '123456' }
+  })
+
+  describe('unauthenticated', () => {
+    beforeEach(() => {
+      authenticatedUser = null
+    })
+
+    it('throws AuthorizationError', async () => {
+      await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+        data: { VerifyEmailAddress: null },
+        errors: [{ message: 'Not Authorised!' }],
+      })
+    })
+  })
+
+  describe('authenticated', () => {
+    beforeEach(async () => {
+      user = await factory.create('User', { email: 'user@example.org' })
+      authenticatedUser = await user.toJson()
+    })
+
+    describe('if no unverified `EmailAddress` node exists', () => {
+      it('throws UserInputError', async () => {
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: { VerifyEmailAddress: null },
+          errors: [{ message: 'Invalid nonce or no email address found.' }],
+        })
+      })
+    })
+
+    describe('given a `EmailAddressRequest`', () => {
+      let emailAddress
+      beforeEach(async () => {
+        emailAddress = await factory.create('EmailAddressRequest', {
+          nonce: 'abcdef',
+          verifiedAt: null,
+          createdAt: new Date().toISOString(),
+          email: 'to-be-verified@example.org',
+        })
+      })
+
+      describe('given invalid nonce', () => {
+        it('throws UserInputError', async () => {
+          variables.nonce = 'asdfgh'
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: { VerifyEmailAddress: null },
+            errors: [{ message: 'Invalid nonce or no email address found.' }],
+          })
+        })
+      })
+
+      describe('given valid nonce for `EmailAddressRequest` node', () => {
+        beforeEach(() => {
+          variables = { ...variables, nonce: 'abcdef' }
+        })
+
+        describe('but the address does not belong to the authenticated user', () => {
+          it('throws UserInputError', async () => {
+            await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+              data: { VerifyEmailAddress: null },
+              errors: [{ message: 'Invalid nonce or no email address found.' }],
+            })
+          })
+        })
+
+        describe('and the `EmailAddressRequest` belongs to the authenticated user', () => {
+          beforeEach(async () => {
+            await emailAddress.relateTo(user, 'belongsTo')
+          })
+
+          it('adds `verifiedAt`', async () => {
+            await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+              data: {
+                VerifyEmailAddress: {
+                  email: 'to-be-verified@example.org',
+                  verifiedAt: expect.any(String),
+                  createdAt: expect.any(String),
+                },
+              },
+              errors: undefined,
+            })
+          })
+
+          it('connects the new `EmailAddress` as PRIMARY', async () => {
+            await mutate({ mutation, variables })
+            const result = await neode.cypher(`
+            MATCH(u:User)-[:PRIMARY_EMAIL]->(e:EmailAddress {email: "to-be-verified@example.org"})
+            MATCH(u:User)<-[:BELONGS_TO]-(:EmailAddress {email: "user@example.org"})
+            RETURN e
+          `)
+            const email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email.toJson()).resolves.toMatchObject({
+              email: 'to-be-verified@example.org',
+            })
+          })
+
+          it('removes previous PRIMARY relationship', async () => {
+            const cypherStatement = `
+            MATCH(u:User)-[:PRIMARY_EMAIL]->(e:EmailAddress {email: "user@example.org"})
+            RETURN e
+          `
+            let result = await neode.cypher(cypherStatement)
+            let email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email.toJson()).resolves.toMatchObject({
+              email: 'user@example.org',
+            })
+            await mutate({ mutation, variables })
+            result = await neode.cypher(cypherStatement)
+            email = neode.hydrateFirst(result, 'e', neode.model('EmailAddress'))
+            await expect(email).toBe(false)
+          })
+
+          describe('Edge case: In the meantime someone created an `EmailAddress` node with the given email', () => {
+            beforeEach(async () => {
+              await factory.create('EmailAddress', { email: 'to-be-verified@example.org' })
+            })
+
+            it('throws UserInputError because of unique constraints', async () => {
+              await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+                data: { VerifyEmailAddress: null },
+                errors: [{ message: 'A user account with this email already exists.' }],
+              })
+            })
+          })
+        })
+      })
+    })
+  })
+})
```


Wow, this is thoroughly tested! 👍

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T09:35:27Z" title="Wednesday, October 2nd 2019, 11:35:27 am +02:00">Oct 2, 2019</time>_

---


I love it!! I guess you probably didn't Test Drive this since the implementation changed a few times??

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-02T13:29:55Z" title="Wednesday, October 2nd 2019, 3:29:55 pm +02:00">Oct 2, 2019</time>_

---


Hard to believe, but I test drove the backend :wink: 

For the frontend I added tests after the implementation because at the time I had hopes to find a solution with less page components and redirects. In the end I figured that solution to be the least complex though.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T14:35:41Z" title="Wednesday, October 2nd 2019, 4:35:41 pm +02:00">Oct 2, 2019</time>_

---


awesome :clap: